Sming/SmingCore/Data/ObjectMap.h

```diff
@@ -0,0 +1,271 @@
+ * Created 2015 by Skurydin Alexey
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * ObjectMap.h
+ *
+ * @author: 31 Jul 2018 - Mikee47 <mike@sillyhouse.net>
+ *
+ */
+
+#ifndef _SMING_CORE_DATA_OBJECT_MAP_H_
+#define _SMING_CORE_DATA_OBJECT_MAP_H_
+
+#include "WVector.h"
+
+/** @brief Implementation of a HashMap for owned objects, i.e. anything created with new().
+ *  @note Once added to the map the object is destroyed when no longer required.
+ *
+ *  To free an object, use one of:
+ *
+ *  ```
+ *  map.remove(key);
+ * 	map.removeAt(index);
+ * 	map[key] = nullptr; // Free existing object and set to null
+ * 	```
+ */
+template <typename K, typename V> class ObjectMap
+{
+public:
+	ObjectMap()
+	{
+	}
+
+	~ObjectMap()
+	{
+		clear();
+	}
+
+	/* Allows operator[] to be used to safely set values */
+	class Value
+	{
+	public:
+		/* Functor to provide guarded access to values */
+		Value(V*& value) : value(value)
+		{
+		}
+
+		Value& operator=(V* newValue)
+		{
+			delete value;
+			value = newValue;
+			return *this;
+		}
+
+		operator const V*() const
+		{
+			return value;
+		}
+
+		operator V*()
+		{
+			return value;
+		}
+
+		V* operator->()
+		{
+			return value;
+		}
+
+	private:
+		V*& value;
+	};
+
+	/**
+	 * @brief Get the number of entries in this map
+	 * @retval int Entry count
+	 */
+	unsigned count() const
+	{
+		return entries.count();
+	}
+
+	/*
+	 * @brief Get a key at a specified index, non-modifiable
+	 * @param idx the index to get the key at
+	 * @return The key at index idx
+	 */
+	const K& keyAt(unsigned idx) const
+	{
+		return entries[idx].key;
+	}
+
+	/*
+	 * @brief Get a key at a specified index
+	 * @param idx the index to get the key at
+	 * @return Reference to the key at index idx
+	 */
+	K& keyAt(unsigned idx)
+	{
+		return entries[idx].key;
+	}
+
+	/*
+	 * @brief Get a value at a specified index, non-modifiable
+	 * @param idx the index to get the value at
+	 * @retval The value at index idx
+	 * @note The caller must not use `delete` on the returned value
+	 */
+	const V* valueAt(unsigned idx) const
+	{
+		return entries[idx].value;
+	}
+
+	/*
+	 * @brief Get a value at a specified index
+	 * @param idx the index to get the value at
+	 * @retval The value at index idx
+	 * @note Because a reference is returned any existing value must be `delete`d first
+	 * @see `operator[]`
+	 */
+	Value valueAt(unsigned idx)
+	{
+		return entries[idx].value;
+	}
+
+	/**
+	 * @brief Get value for given key, if it exists
+	 * @param key
+	 * @retval const V* Will be null if not found in the map
+	 */
+	const V* operator[](const K& key) const
+	{
+		return find(key);
+	}
+
+	/** @brief Access map entry by reference
+	 *  @param key
+	 *  @retval Value Guarded access to mapped value corresponding to given key
+	 *  @note If the given key does not exist in the map then it will be created and a null value entry returned.

```

Comment:
To avoid this (usually) un-desirable side-effect we could take advantage of our `Value` functor and use it to create an entry in the map only when a value is actually assigned.